Introduse the ability to postpone a downloaded update #4263

LeonarddeR opened the pull request:
Closes #4263
closes #6281
closes #4020

First of all, many thanks to @zahyur  who actually did the major work for this.
This pull request introduces the possibility to postpone a downloaded update. Comment and discussion in #4263 has been processed in this request. Also, I decided to make installing postponed updates from the menu possible regardless of showing the exit dialog has been disabled.

This code still needs testing in real live, which is quite hard from source, without real updatable builds.

### Changelog entry
* New Features
    + When updating NVDA, it is now possible to postpone the installation of the update to a later moment of choice. (#4263)
       - When there is an update pending, installation can be performed from the NVDA menu or the NVDA exit dialog.
    + It is now possible to automatically update a portable copy of NVDA when you have write access to the folder where the portable copy resides. (#6281)

zahyur committed:
initial work on postponeing updates

Commit:
Removed undesired code from the source and userdocs and added some copyright notices

Commit:
Merge remote-tracking branch 'origin/master' into t4263

Commit:
Merge remote-tracking branch 'origin/master' into t4263, fixing merge conflicts

leonardder commented:
I wonder, is there still desire to leave this open? I initially found this a quite interesting idea and the code was already there for the most part.

cc @feerrenrut

feerrenrut commented:
Yes, I think so. I will review it shortly. Unless it has been obsoleted by some other change?

feerrenrut commented on source/updateCheck.py:
```diff
@@ -242,17 +283,78 @@ def onDownloadButton(self, evt):
+			message=_("You can install it now, or install it later from the NVDA menu or the exit dialog, if enabled.\n")
+		mainSizer.Add(wx.StaticText(self, label=
+			# Translators: A message indicating that an updated version of NVDA is ready to be installed.
+			_("Update is ready to install.\n")+message+
```
This building up of messages has a risk that it won't make sense once translated. I also think these labels could be simplified.  I think it is adequate to simply set the label as "Update is ready to install" and rely on the presence of the buttons to indicate what options are available.

feerrenrut commented on source/updateCheck.py:
```diff
@@ -242,17 +283,78 @@ def onDownloadButton(self, evt):
+		except:
+			gui.messageBox(
+				# Translators: The message when a downloaded update file could not be preserved.
+				_("The downloaded file could not be postponed properly."),
```
I would word this "Unable to postpone update"

feerrenrut commented on source/updateCheck.py:
```diff
@@ -242,17 +283,78 @@ def onDownloadButton(self, evt):
+			gui.messageBox(
+				# Translators: The message when a downloaded update file could not be preserved.
+				_("The downloaded file could not be postponed properly."),
+				_("Error"), wx.OK | wx.ICON_ERROR)
```
This is missing a translator comment. I think there is an automated test that would catch this, you can run this with `scons checkPot`

feerrenrut reviewed (changes requested):
Sorry it's taken a while for this to be looked at. I think I will need to go over this again, but hopefully we can start by agreeing on the UX for this change.

Commit:
Merge remote-tracking branch 'origin/master' into t4263

Commit:
Review actions

leonardder commented:
It seems that when I reviewed this code earlier and made it ready for a pr, I missed some major things that really need to be fixed before this even works. The major downside with this code is that it is very difficult to test, at least when running from source. @feerrenrut, I've addressed your review comments although not yet pushed. However, let me do some additional testing before I ask you to come back to this again.

Commit:
Some major code updates to make this actually work, hopefully

Commit:
Fixed translator comment

Commit:
Some more fixes, still some testing to be done

leonardder commented:
I've been able to test this to some extend, however as pointed out, it requires some ugly hacking as NVDA only allows itself to be updated in official snapshot or release cases. I'd also like to incorporate support for portable copy updating while at it.

Commit:
Support updating of portable copies

feerrenrut commented:
Ok, please request a review again when you are ready for me to take another look at it.

Commit:
Merge remote-tracking branch 'origin/master' into t4263, fixing merge conflicts.

feerrenrut commented on source/gui/__init__.py:
```diff
@@ -200,6 +200,18 @@ def onTemporaryDictionaryCommand(self,evt):
+		try:
+			self.sysTrayIcon.menu.RemoveItem(self.sysTrayIcon.runPendingUpdateMenuItem)
+		except:
+			pass
```
I think it's worth logging something here at debug level.

feerrenrut commented on source/gui/__init__.py:
```diff
@@ -200,6 +200,18 @@ def onTemporaryDictionaryCommand(self,evt):
+		except:
+			pass
+		if updateCheck and updateCheck.isPendingUpdate():
+			self.sysTrayIcon.menu.InsertItem(self.sysTrayIcon.menu.GetMenuItemCount()-2,self.sysTrayIcon.runPendingUpdateMenuItem)
```
What's with the `-2`. Is this the position. Might be clearer to put this in a `pendingUpdateMenuItemPosition` variable. Is it important for it to be second from the bottom? I am concerned that because this positioning is separated from the rest of that menu filing that the order might change, and this code wont be considered. The index of the position for this item could be saved at the point where the rest of the menu is populated.

feerrenrut commented on source/config/__init__.py:
```diff
@@ -250,6 +250,8 @@ def _setSystemConfig(fromPath):
 		if not os.path.isdir(curDestDir):
 			os.makedirs(curDestDir)
 		for f in files:
+			if f.endswith(".exe"):
```
perhaps a comment to say why exe files should be skipped?

feerrenrut commented on source/updateCheck.py:
```diff
@@ -98,6 +106,53 @@ def checkForUpdate(auto=False):
+	state["removeFile"] = destPath
+	saveState()
+	if config.isInstalledCopy():
+		exeCuteParams = u"--install -m"
```
this looks like a bug, here `exeCuteParams` has a capital C, and where it is passed to ShellExecute it does not.

feerrenrut commented on source/updateCheck.py:
```diff
@@ -245,32 +308,79 @@ def onDownloadButton(self, evt):
 		saveState()
 		self.Close()
 
+class UpdateAskInstallDialog(wx.Dialog):
```
I have a feeling that this dialog wont have the visual spacing expected between labels and buttons. What is the easiest way to test this, so I can inspect it?

leonardder replied:
I added some guiHelper stuff to it now. Here is a code snippet to make it show up:

```
import versionInfo
versionInfo.updateVersionType="snapshot:next"
import updateCheck
updateCheck.initialize()
updateCheck.UpdateAskInstallDialog(gui.mainFrame, "dummy_path", "2018.3")
```

feerrenrut replied:
Thanks for the snippet. It looks ok to me.

feerrenrut commented on source/updateCheck.py:
```diff
@@ -390,19 +500,9 @@ def _error(self):
 			wx.OK | wx.ICON_ERROR)
 
 	def _downloadSuccess(self):
+#		global storeUpdatesDir
```
Looks like this can be removed?

Commit:
Merge remote-tracking branch 'origin/master' into t4263

Commit:
Review actions

feerrenrut commented on source/updateCheck.py:
```diff
@@ -500,7 +504,6 @@ def _error(self):
 			wx.OK | wx.ICON_ERROR)
 
 	def _downloadSuccess(self):
-#		global storeUpdatesDir
```
Should this be uncommented / removed?

leonardder replied:
This has already been removed. Probably legacy code from the initial author, but storeUpdatesDir doesn't have to be changed inside this function.

feerrenrut replied:
Oh sorry I must have misread this line.

feerrenrut commented on source/gui/__init__.py:
```diff
@@ -773,14 +794,18 @@ def __init__(self, parent):
+		]
+		if updateCheck and updateCheck.isPendingUpdate():
+			# Translators: An option in the combo box to choose exit action.
+			self.actions.append(_("Run pending update"))
```
Elsewhere we say "Install update" rather than "run update", I think "Install pending update" would be more consistent.

feerrenrut commented on source/updateCheck.py:
```diff
@@ -245,32 +307,84 @@ def onDownloadButton(self, evt):
+		sHelper = guiHelper.BoxSizerHelper(self, orientation=wx.VERTICAL)
+		
+		# Translators: A message indicating that an updated version of NVDA is ready to be installed.
+		sHelper.addItem(wx.StaticText(self, label=_("Update is ready to install.\n")))
```
Do you think this dialog should tell you the updates version information? To be consistent with the other download update dialog?

Commit:
More review actions

Commit:
Actions for user guide

leonardder commented:
@feerrenrut, just to make sure, is it clear to you that this pr also addresses #6281?

This pr requires extensive testing. May be, just for the sake of stability, it is a good idea not to target 2018.1 for this. It is most likely that there are some edge cases we missed.

feerrenrut commented:
Yes, I think that letting it hang around for the rest of the dev cycle for 2018.2 is best. Thanks for updating the description re #6281.

gdata1 commented:
Hello. I'm downloading an update, but I can't install anything else. NVDA next-14821,51682706.

gdata1 commented:
INFO - __main__ (07:20:53.002):
Starting NVDA
INFO - core.main (07:20:53.142):
Config dir: C:\Users\nikit\AppData\Roaming\nvda
INFO - config.ConfigManager._loadConfig (07:20:53.142):
Loading config: C:\Users\nikit\AppData\Roaming\nvda\nvda.ini
INFO - core.main (07:20:53.266):
NVDA version next-14816,6626ffe4
INFO - core.main (07:20:53.266):
Using Windows version 10.0.17083 workstation
INFO - core.main (07:20:53.266):
Using Python version 2.7.14 (v2.7.14:84471935ed, Sep 16 2017, 20:19:30) [MSC v.1500 32 bit (Intel)]
INFO - core.main (07:20:53.266):
Using comtypes version 1.1.3
INFO - external:synthDrivers.RHVoice.SynthDriver.__init__ (07:20:53.477):
Using RHVoice version 0.5
INFO - synthDriverHandler.setSynth (07:20:53.482):
Loaded synthDriver RHVoice
INFO - core.main (07:20:53.482):
Using wx version 3.0.2.0 msw (classic)
INFO - brailleInput.initialize (07:20:53.483):
Braille input initialized
INFO - braille.initialize (07:20:53.483):
Using liblouis version 3.3.0
INFO - braille.BrailleHandler.setDisplayByName (07:20:53.484):
Loaded braille display driver noBraille, current display has 0 cells.
WARNING - core.main (07:20:53.516):
Java Access Bridge not available
INFO - _UIAHandler.UIAHandler.MTAThreadFunc (07:20:53.520):
UIAutomation: IUIAutomation3
INFO - core.main (07:20:53.867):
NVDA initialized
INFO - updateCheck.AutoUpdateChecker._started (07:20:53.882):
Performing automatic update check
ERROR - unhandled exception (07:21:16.573):
Traceback (most recent call last):
  File "wx\_core.pyc", line 16766, in <lambda>
  File "updateCheck.pyc", line 330, in __init__
AttributeError: 'StaticText' object has no attribute 'format'

PratikP1 commented:
I’m Getting the update several times. From the log, it appears that NVDA is checking for an update twice. Once downloaded, nothing happens. The update is not installed. When starting NVDA, the check happens again with the checkbox in the general preferences dialog checked. Installation does not occur once again. I can provide the log if necessary.

  *   next-14816,6626ffe4

Commit:
Fixed issue where a right paren was placed at the wrong position

leonardder commented:
@feerrenrut or @michaeldcurran, could you please incubate the most recent change? I'm afraid that people can't update their next versions now. It was actually a quite silly issue which should now be fixed.

Commit:
Properly evaluate whether to show or hide the pending updates menu item on every time the menu is opened. Binding this evaluation to the item itself made the item never to show up.

Commit:
Don't use the file variable name when looping through possible old updates. This broke the loading of the state file. Added state file loading debug logging while at it.

leonardder commented:
Just pushed two additional fixes:

1. There was an issue in updateCheck.initialize that broke the initial loading of the update state file. This made a pending update go off the radar. Never use file as a variable name inside a loop! :)
2. Whether or not to show the install pending update menu item was only evaluated when that item was in the menu. This meant that this item would never show up properly, unless you'd restart NVDA after postponing an update.

k-kolev1985 commented:
Maybe a bug to report. I've received the notification for a new "next" snapshot update today. In order to see how the update postpone feature is implemented, I've decided to postpone it. After that, I've invoked it from the "Check for updates" command in NVDA menu -> Help -> Check for updates. It asked me if I want to install the pending update or to redownload it. I chose to install it. The "Please wait while setup is loading" dialog appeared, but the previous one that asked me what to do did not disappear, until the current copy of NVDA was shut down. Isn't it supposed to disappear when the chosen command is activated?

Commit:
Properly destroy the UpdateResultDialog when installing a postponed update

leonardder commented:
@k-kolev1985, thanks for your report, I just commited a fix for this. I hope @feerrenrut will be able to come back to this pr next Monday.

Commit:
When updating a portable copy, run NVDA with the -m command line argument

feerrenrut commented on source/gui/__init__.py:
```diff
@@ -474,11 +488,17 @@ def __init__(self, frame):
 			# Translators: The label for the menu item to save current settings.
 			item = self.menu.Append(wx.ID_SAVE, _("&Save configuration"), _("Write the current configuration to nvda.ini"))
 			self.Bind(wx.EVT_MENU, frame.onSaveConfigurationCommand, item)
-		if not globalVars.appArgs.secure:
```
Removing this line seems like a regression to me?

leonardder replied:
I should have pointed this out beforehand. What's not clear from the diff is that the code part below the removed line is also part of an indented block for exactly the same if check. So this particular if check can be safely removed to make this code part of the previous if block, which is equivalent. It is quite unrelated to this pr though.

feerrenrut replied:
Oh yep

feerrenrut commented:
I have an installed version of `next-14810,33118eb1` and I can not update to the latest next. After clicking OK to "Update has been downloaded, it will now be installed", I hear a windows error sound, and get the following in the log:
```
DEBUG - core.main (16:56:38.171):
initializing updateCheck
DEBUGWARNING - watchdog._watcher (16:56:38.674):
Trying to recover from freeze, core stack:
  File "nvda.pyw", line 214, in <module>
  File "logging\__init__.pyc", line 1212, in critical
  File "logHandler.pyc", line 142, in _log
  File "logging\__init__.pyc", line 1286, in _log
  File "logging\__init__.pyc", line 1296, in handle
  File "logging\__init__.pyc", line 1336, in callHandlers
  File "logHandler.pyc", line 225, in handle

CRITICAL - __main__ (16:56:38.171):
core failure
Traceback (most recent call last):
  File "nvda.pyw", line 212, in <module>
  File "core.pyc", line 469, in main
  File "updateCheck.pyc", line 586, in initialize
KeyError: 'pendingUpdateVersion'
DEBUGWARNING - RPC process 11988 (nvda_update_fxyzrv.exe) (16:56:45.990):
Thread 1860, build\x86\remote\ia2Support.cpp, IA2Support_inProcess_terminate, 152:
WaitForMultipleObjects returned 258
```

Commit:
Fix KeyError on updating

leonardder commented:
It turns out that the previous bug with the state file obfuscated 
another bug, getting a key from a dictionary which did not exist yet. 
This should have been fixed.

Commit:
Merge branch 't4263' of github.com:/babbagecom/nvda into t4263

Commit:
make sure that the update state file will be updated to contain pending update info

Commit:
When updating a portable copy, place paths that are provided as command line params between quotes to deal with paths containing spaces

Brian1Gaff commented:
Interesting that. That was what I had back when I did updates before and it 
failed to restart as the log was not writable so it chucked it in the nvda 
folder as a separate log.
 I'm sure you remember. Now I get around this inability to cope by lots of 
pauses wherever there are buttons or dialogues as windows often gets busy 
behind the scenes and seemingly has not finished what its doing with some 
files it needs to terminate or free up.
 Brian

Commit:
Fix crash after restart with a postponed update

leonardder commented:
@feerrenrut: Just committed an additional fix. The startup notification for pending updates will now be handled by the AutoUpdateChecker instead of an askInstallUpdateNow method on updateCheck which did not exist. :P

@MichaelDCurran: May be it's a good idea if you could also do a quick review of the code in case there's another relevant thing we missed.

leonardder commented:
Note that with my most recent change, the behavior of the "Notify for pending update on startup" option has changed a little. As this functionality now resides in the AutoUpdateChecker, a pending update will be ignored if there is a version available that is newer than the pending update. I think this behavior is desirable, it does not make sense to install a pending update and notice afterwards that it wasn't the most recent update available. As a side effect, the pending update notification won't show up if the update check fails due to no internet connection available. People will still be able to install the postponed update from the menu or using the exit dialog, though.

Brian1Gaff commented:
The latest next installed as portable OK this time, though the  dialogue did 
not gain focus on its own, I had to alt tab to it.
 Not tried defer update yet.
 Brian

Commit:
No longer add removeFile to the update state, as it is used nowhere

Commit:
Call prePopup before raising the UpdateAskInstallDialog and make it modal

feerrenrut commented on source/updateCheck.py:
```diff
@@ -352,7 +354,8 @@ def __init__(self, parent, destPath, version):
 		mainSizer.Fit(self)
 		self.Center(wx.BOTH | wx.CENTER_ON_SCREEN)
-		self.Show()
+		gui.mainFrame.prePopup()
```
what's this about?

feerrenrut replied:
If you call this, I think you also need to call `postPopup()`.

leonardder replied:
Yes, you'd think so. The use of prePopup and postPopup inside updateCheck really puzzles me.

1. prePopup is only used in the UpdateDownloader, not for the other dialogs. It was called for the old update installation message box, so that's why I belief it should be called in the UpdateAskInstallDialog as well.
2. postPopup is even more confusing given it is being called in a wx.CallLater with a not very descriptive comment. A git blame points at @jcsteh, so may be he knows more?
3. In several places in the gui module, you see the following code:

    ```
mainFrame.prePopup()
dialog.Show()
mainFrame.postPopup()
```

    However, this doesn't seem to make much sense, as Show doesn't block the caller. For ShowModal, it does make sense as ShowModal blocks the caller until the dialog is dismissed.

feerrenrut replied:
Perhaps its better to use `def runScriptModalDialog` in `source/gui/__init__`

jcsteh replied:
TL;DR: The general rule is that if you're popping up a window (as opposed to a window which appears as a result of some action performed while our GUI was already visible), you must call `prePopup`. If you call `prePopup`, there must be a corresponding call to `postPopup`. Ideally, this `postPopup` call would occur once the window is dismissed, but for windows that aren't modal, it's enough in most cases to just call it after `Show`.

`prePopup` and `postPopup` have two major functions:

1. To force Windows to bring us to the foreground (`prePopup`) and to return to the previous foreground (`postPopup`). The way foreground stuff works in Windows is really obscure. `prePopup` mostly makes sense; we ask Windows to bring us to the foreground. However, we were seeing cases where the user would just be stuck in limbo when our pop-ups were dismissed. I still don't really understand why this is the case, but the show/hide in `postPopup` seemed to fix it. (Obviously, that doesn't actually get applied at the right tie for dialogs that aren't modal, but it seems to behave regardless.) This code is very old, wx has been updated and Windows XP is no longer a factor, so you could consider re-assessing to see whether we can get rid of this ugliness. However, note that installed versus portable is still a factor, since uiAccess gives the app special privileges regarding the foreground window.
2. These methods also set things up so that NVDA is aware of what the focus was before the pop-up. This is important for input gestures, for example, so it knows which object to get gestures for. `postPopup` does the necessary teardown of this state. The fact that it gets called immediately after `Show` is fine in this case because the dialog should already have done what it needs to do with this state.

leonardder commented:
The prePopup should fix the issue reported by Brian about this dialog not getting properly focused. I'm still not very sure about when to use prePopup and postPopup.

> Op 9 feb. 2018 om 07:35 heeft Reef Turner <notifications@github.com> het volgende geschreven:
> 
> @feerrenrut commented on this pull request.
> 
> In source/updateCheck.py:
> 
> > @@ -352,7 +354,8 @@ def __init__(self, parent, destPath, version):
>  		self.Sizer = mainSizer
>  		mainSizer.Fit(self)
>  		self.Center(wx.BOTH | wx.CENTER_ON_SCREEN)
> -		self.Show()
> +		gui.mainFrame.prePopup()
> what's this about?
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

Commit:
Use runScriptModalDialog for the UpdateAskInstallDialog

leonardder commented:
@feerrenrut: IN https://github.com/nvaccess/nvda/pull/6355/commits/20407fba208cfca37e7606959d3b0d08b2ead8da I started to use scriptRunModalDialog

feerrenrut commented:
Ok, looks good. I will incubate this again.

Commit:
Clear the don't remind version when postponing an update. If we do not do this, a user will never be reminded about a postponed update automatically at all

leonardder commented:
@feerrenrut: Three cheers for broken code. Just tried to fix another small bug, namely that postponing an update does not clear the don't remind version. This means that the don't remind version and the pending version are equal, in which case the AutoUpdateChecker would never pop up a reminder.

leonardder commented:
@feerrenrut: Three cheers for broken code. Just tried to fix another small bug, namely that postponing an update does not clear the don't remind version. This means that the don't remind version and the pending version are equal, in which case the AutoUpdateChecker would never pop up a reminder.

Brian1Gaff commented:
just ran the portable version of next and had to prompt it to tell me a new 
version was ready, but once downloaded, this time it did show me the install 
and postpone box automatically which did not happen before.
 Brian

leonardder commented:
@feerrenrut: Could you please review 4a4f7c4d157e12b3e030f0b402997380769f7884?

feerrenrut reviewed (approved):
The latest commit makes sense. I think it would be handy to have a manual test plan for this feature. Outline all the possible paths that one can take, and what is expected to happen. It might help to highlight some case that has not yet been explored.
